### ISSUE TITLE
Pupil Planes can be too big for galsim.optics

### ISSUE THREAD

**Issue by cpadavis** [GalSim-developers/GalSim#752]
In galsim.optics you can put in a highres pupil image and the code will fail, whereas if you deres the image (say, take every 4 of it), the code will run.

```python
import galsim.optics
from astropy.io import fits
resample_factor = 1
pupil_plane_data = fits.getdata('DECam_pupil_512.fits')
# the pupil is oversampled relative to PSF by 8
if resample_factor > 1:
    pupil_plane_data = pupil_plane_data[::resample_factor, ::resample_factor]
pupil_plane_im = galsim.Image(pupil_plane_data)
opt = galsim.optics.OpticalPSF(aberrations=[0,0,0,0,0.5,0.1], diam=4, lam=700, pupil_plane_im=pupil_plane_im, oversampling=int(8 / resample_factor))
```

Which if you run, you get:

```python
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-8-30326ca7dfa5> in <module>()
      8     pupil_plane_data = pupil_plane_data[::resample_factor, ::resample_factor]
      9 pupil_plane_im = galsim.Image(pupil_plane_data)
---> 10 opt = galsim.optics.OpticalPSF(aberrations=[0,0,0,0,0.5,0.1], diam=4, lam=700, pupil_plane_im=pupil_plane_im, oversampling=oversampling)

/u/ki/cpd/.local/lib/python2.7/site-packages/galsim/optics.pyc in __init__(self, lam_over_diam, lam, diam, defocus, astig1, astig2, coma1, coma2, trefoil1, trefoil2, spher, aberrati
ons, circular_pupil, obscuration, interpolant, oversampling, pad_factor, suppress_warning, max_size, flux, nstruts, strut_thick, strut_angle, pupil_plane_im, pupil_angle, scale_unit
, gsparams)
    331             aberrations=aberrations, circular_pupil=circular_pupil, obscuration=obscuration,
    332             flux=flux, nstruts=nstruts, strut_thick=strut_thick, strut_angle=strut_angle,
--> 333             pupil_plane_im=pupil_plane_im, pupil_angle=pupil_angle, oversampling=oversampling)
    334
    335         # Initialize the GSObject (InterpolatedImage)

/u/ki/cpd/.local/lib/python2.7/site-packages/galsim/optics.pyc in psf_image(array_shape, scale, lam_over_diam, aberrations, circular_pupil, obscuration, nstruts, strut_thick, strut_
angle, flux, pupil_plane_im, pupil_angle, oversampling)
    933         if oversamp_ratio < 0.9:
    934             raise RuntimeError("Supplied pupil plane image results in inadequate sampling "
--> 935                                "by factor of %f"%oversamp_ratio)
    936         im = galsim.Image(array.astype(np.float64), scale=scale/oversamp_ratio)
    937     else:

RuntimeError: Supplied pupil plane image results in inadequate sampling by factor of 0.555515
```

Now, if you resample the image from 512x512 to 128x128, it passes! Also, changing the oversampling parameter doesn't change the error.

**Comment by jmeyers314** [comment]
Thanks for pointing this out @cpadavis .  After some investigation, I think I understand why this is happening.  The size of the pupil plane array is set (from the diameter, wavelength, obscuration, pad_factor, and oversampling params) before the routine even knows whether the pupil is defined via geometric parameters or from an image.  In this case, with oversampling=8, it picks 2048x2048:

```
import galsim
airy = galsim.Airy(diam=4, lam=700)
oversampling = 8
pad_factor = 1.5
print galsim._galsim.goodFFTSize(int(2 * airy.maxK() / airy.stepK() * oversampling * pad_factor))
# 2048
```

The supplied image in this case is only 512x512, so it gets zero padded up to 2048x2048.  The original image is illuminated up to pixel number ~230 (where 0 means the image center).  Nyquist sampling (i.e., oversampling=1.0) happens when the size of the image is twice the size of the illuminated region, so the effective oversampling is 2048 / (2 * 460) = 2.22, which is less than the requested 8, and the code yells at you.

When the image is thinned down to 128x128, though, it still gets padded up to 2048x2048 (remember, that was decided before the code even knew there would be an image).  Now the effective oversampling is 2048 / (2 * 460/4) = 8.88 > 8 and the code is happy.  What's been gained in PSF sampling gets lost in the size of the PSF array, though (and aliasing has also increased).

Probably the API would be more intuitive if instead of setting a pupil size and scale before looking at the input image, the scale was inferred from the image (either directly from the image.scale attribute, or from the farthest illuminated pixel + the diameter keyword), and then the pupil size was set from the oversampling keyword _after_ the scale determination.  To handle pad_factor, it might be necessary to resample the pupil plane image to higher resolution too.  I'll work on implementing these conventions in the new API (they're not there yet).

**Comment by jmeyers314** [comment]
Hi @cpadavis, I think this is now fixed on branch #549, which should be merged into master soon.  The basic change in strategy is to use the diam kwarg to decide a good pupil plane size in meters (instead of pixels), and then once the image has been processed (including determining the pixel scale), pad if necessary.

If you want to verify that everything makes sense, please do checkout that branch and let me know.  You might take a look at test_optics.py:test_OpticalPSF_pupil_plane_size() in particular, which I believe is equivalent to your example above, reproduces your error under the old optics framework, and passes under the new framework.

**Comment by jmeyers314** [comment]
@cpadavis , can you checkout the latest version of the master branch and see if we can close this issue?  Thanks!

**Comment by cpadavis** [comment]
I just want to make sure of this: galsim.optics.OpticalPSF can still be imported but is the old code (and so will still yield this error). But if I import galsim.OpticalPSF, I have the new code and everything works.

with galsim.OpticalPSF (like in your example), it works!

**Comment by rmjarvis** [comment]
Good point.  Yes, your example does have `galsim.optics.OpticalPSF`.  The `optics.` had previously been unnecessary, but allowed.  Now that should give you a deprecation warning and continue to give the error.  But `galsim.OpticalPSF` should work.

**Comment by jmeyers314** [comment]
>, it works!

Great!  Closing this issue.  (Fixed on #549 ).
